integrations/jira/jira/client.py

```diff
@@ -54,11 +52,103 @@ def __init__(self, jira_url: str, jira_email: str, jira_token: str) -> None:
             self.jira_api_auth = BasicAuth(self.jira_email, self.jira_token)
 
         self.api_url = f"{self.jira_rest_url}/api/3"
+        self.teams_base_url = f"{self.jira_url}/gateway/api/public/teams/v1/org"
         self.webhooks_url = f"{self.jira_rest_url}/webhooks/1.0/webhook"
 
         self.client = http_async_client
         self.client.auth = self.jira_api_auth
         self.client.timeout = Timeout(30)
+        self.semaphore = asyncio.Semaphore(MAX_CONCURRENT_REQUESTS)
+
+    async def _send_api_request(
+        self,
+        method: str,
+        url: str,
+        params: Optional[dict[str, Any]] = None,
+        json: Optional[dict[str, Any]] = None,
+        headers: Optional[dict[str, str]] = None,
+    ) -> Any:
+        async with self.semaphore:
+            response = await self.client.request(
+                method=method, url=url, params=params, json=json, headers=headers
+            )
+            response.raise_for_status()
+            return response.json()
+
+    async def _get_paginated_data(
+        self,
+        url: str,
+        extract_key: Optional[str] = None,
+        page_size: int = PAGE_SIZE,
+        initial_params: Optional[dict[str, Any]] = None,
+    ) -> AsyncGenerator[list[dict[str, Any]], None]:
+        params = initial_params or {}
+        params.update(self._generate_base_req_params())
+        params["maxResults"] = page_size
+
+        start_at = 0
+        while True:
+            params["startAt"] = start_at
+            response_data = await self._send_api_request("GET", url, params=params)
+
+            if extract_key:
+                items = response_data.get(extract_key, [])
+            else:
+                items = (
```

Comment:
What methods require this block?

Reply:
If `get_paginated_users` is the only one that requires a list, it shouldn't be calling this endpoint since only one call is sufficient to retrieve the users. It should call the `_send_api_request` method directly.

When you introduce "special cases" to an algorithm, you increase its complexity and make it more difficult to use. Let's leave this method to handle only real offset based pagination.